### ISSUE TITLE
redundant_closure_call breaks code

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9956]
### Summary

.

### Lint Name

redundant_closure_call

### Reproducer

I tried this code:

```rust
fn main() {
    assert_eq!((||||42)()(), 42);
}
```

I saw this happen:
`cargo clippy --fix`
```
  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0618]: expected function, found `{integer}`
 --> src/main.rs:2:18
  |
2 |     assert_eq!(||42(), 42);
  |                  ^^--
  |                  |
  |                  call expression requires function

error: aborting due to previous error

For more information about this error, try `rustc --explain E0618`.
Original diagnostics will follow.

For more information about this error, try `rustc --explain E0618`.
error: could not compile `fff` due to previous error
warning: build failed, waiting for other jobs to finish...
warning: try not to call a closure in the expression where it is declared
 --> src/main.rs:2:16
  |
2 |     assert_eq!((||||42)()(), 42);
  |                ^^^^^^^^^^ help: try doing something like: `||42`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure_call
  = note: `#[warn(clippy::redundant_closure_call)]` on by d
```

I expected to see this happen:

### Version

```text
rustc 1.67.0-nightly (b3bc6bf31 2022-11-24)
binary: rustc
commit-hash: b3bc6bf31265ac10946a0832092dbcedf9b26805
commit-date: 2022-11-24
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
